src/js/tls.js

```diff
@@ -15,184 +15,137 @@
+  function onclose() {
+    this.emit('close');
   }
 
-  if (opts.SNICallback && !util.isFunction(opts.SNICallback)) {
-    throw new TypeError('TLSSocket - options.SNICallback must be a function.');
+  function onfinish() {
+    this.emit('finish');
   }
 
-  if (opts.session && !util.isBuffer(opts.session)) {
-    throw new TypeError('TLSSocket - options.session should be a buffer.');
+  function onend() {
+    this.emit('end');
   }
-};
 
-Tls.TLSSocket.prototype.address = function() {
-  throw new TypeError('Unimplemented');
-};
+  this._socket.on('close', onclose.bind(this));
+  this._socket.on('finish', onfinish.bind(this));
+  this._socket.on('end', onend.bind(this));
 
-Tls.TLSSocket.prototype.disableRenegotiation = function() {
-  this.disableRenegotiation = true;
-};
-
-Tls.TLSSocket.prototype.encrypted = function() {
-  return this.encrypted;
-};
+  // Native handle
+  this._tls = new native.TlsConstructor(this.options);
+  this._tls.jsref = this;
+  this._tls.onread = this.onread;
+  this._tls.onwrite = this.onwrite;
+  this._tls.onclose = this.onclose;
+  this._tls.onhandshakedone = this.onhandshakedone;
+  this._tls.ondata = this.ondata;
+}
+util.inherits(TLSSocket, EventEmitter);
 
-Tls.TLSSocket.prototype.getCipher = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.connect = function(opts, callback) {
+  this.once('connect', callback);
+  this._socket.connect(opts);
+  return this;
 };
 
-Tls.TLSSocket.prototype.getEphemeralKeyInfo = function() {
-  throw new TypeError('Unimplemented');
-};
+TLSSocket.prototype.write = function(data, callback) {
+  if (!Buffer.isBuffer(data)) {
+    data = new Buffer(data);
+  }
 
-Tls.TLSSocket.prototype.getPeerCertificate = function(/* detailed */) {
-  throw new TypeError('Unimplemented');
+  return this._socket.write(this._tls.write(data), callback);
 };
 
-Tls.TLSSocket.prototype.getProtocol = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.pause = function() {
+  this._socket.pause();
 };
 
-Tls.TLSSocket.prototype.getSession = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.resume = function() {
+  this._socket.resume();
 };
 
-Tls.TLSSocket.prototype.getTLSTicket = function() {
-  throw new TypeError('Unimplemented');
+TLSSocket.prototype.end = function() {
+  this._ended = true;
+  this._socket.end(null, function() {
+    this.end();
+  });
 };
 
-Tls.TLSSocket.prototype.write = function(message) {
-  return native.write(message);
+TLSSocket.prototype.destroy = function() {
+  this._socket.destroy();
 };
 
-Tls.TLSSocket.prototype.renegotiate = function(options, callback) {
-  if (!util.isObject(options)) {
-    throw new TypeError('options should be an object');
-  }
-
-  if (callback && !util.isFunction(callback)) {
-    throw new TypeError('callback should be a function');
+TLSSocket.prototype.onsocket = function() {
+  if (this._ended) {
+    return;
   }
+  this.emit('socket', this._socket);
```

Comment:
Emitting `socket` event doesn't seem a role of tls module. I guess that handshaking tls needed to be started in `onconnection` status.